### PR TITLE
ssh: add ability to enable compression (for SCP/SFTP)

### PR DESCRIPTION
**Pull request by vszakats**
Add the ability to enable compression for SCP/SFTP.

The required low-level logic was already available as part of `libssh2` (via `LIBSSH2_FLAG_COMPRESS` [`libssh2_session_flag()`](https://www.libssh2.org/libssh2_session_flag.html) option.)

This patch adds the new `libcurl` option `CURLOPT_SSH_COMPRESSION` (boolean) and the new `curl` command-line option `--compressed-ssh` to request this `libssh2` feature. To have compression enabled, it is required that the SSH server supports a (zlib) compatible compression method and that `libssh2` was built with `zlib` support enabled.

Ref: https://github.com/curl/curl/issues/1732

---

Use case: When transferring well-compressible (f.e. plaintext) data which is either large and/or the bandwidth is relatively low. For small transfers and/or fast connections, it may even result in slightly worse performance. See also: https://askubuntu.com/questions/6770/does-ssh-use-any-compression

### CODE REVIEW

**Comment by vszakats** [comment]
Travis CI failures:
``` https://travis-ci.org/curl/curl/jobs/261287192#L215
==> ./b2 --prefix=/usr/local/Cellar/boost/1.64.0_1 --libdir=/usr/local/Cellar/bo
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
```
New test 0642 OK, an unrelated one failed:
``` https://travis-ci.org/curl/curl/jobs/261287190#L5488
TESTFAIL: These test cases failed: 1208 
```

**Comment by vszakats** [comment]
I've created the patch assuming it would make it only into 7.56.0. That is to say I'd be quite glad if it could be included in 7.55.0 next week. Given, of course that all eyes find it alright.

**Comment by bagder** [comment]
We only mege features when the feature window is open. That's 4 weeks following a release, so too late for 7.55.0

**Comment by vszakats** [comment]
Ok, it's clearly too late for that, thanks for clarifying!

**Comment by vszakats** [comment]
My real world tests were conclusively positive. Here're are test binaries (for Windows) with this patch applied including a `libssh2` library with `zlib` enabled: https://bintray.com/vszakats/generic/curl-test/7.55.0

(NOTE: these binaries are not for production use or distribution, their sole purpose is testing this specific feature on that platform.)

**Comment by vszakats** [comment]
Any suggestions or anything missing to have this approved?

File: src/tool_operate.c [hunk 1]
```diff
@@ -1091,6 +1091,10 @@ static CURLcode operate_do(struct GlobalConfig *global,
              to fail if we are not talking to who we think we should */
           my_setopt_str(curl, CURLOPT_SSH_HOST_PUBLIC_KEY_MD5,
                         config->hostpubmd5);
+
+          /* new in libcurl 7.56.0 */
+          my_setopt_str(curl, CURLOPT_SSH_COMPRESSION,
```
Comment (jay) [inline src/tool_operate.c]: this should be my_setopt and you have to pass it a long like if(config->ssh_compression) my_setopt(curl, CURLOPT_SSH_COMPRESSION, 1L). Also the example should use 1L

File: tests/data/test642 [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+# Client-side
+<client>
+<server>
+sftp
+</server>
+ <name>
+SFTP retrieval
+ </name>
+ <command>
+--key curl_client_key --pubkey curl_client_key.pub -u %USER: --compressed-ssh sftp://%HOSTIP:%SSHPORT%PWD/log/file642.txt --insecure
```
Comment (jay) [inline tests/data/test642]: the test doesn't do anything to see if the compression is enabled, is that right?

Comment (vszakats) [inline tests/data/test642]: That's right. It tests if download is still (transparently) working.

File: src/tool_operate.c (hunk 1 above)
Comment (vszakats) [inline src/tool_operate.c]: Thanks!

Comment (vszakats) [inline src/tool_operate.c]: Fixed.

File: tests/data/test642 (hunk 2 above)
Comment (vszakats) [inline tests/data/test642]: Fixed.

**Comment by vszakats** [comment]
Squashed and rebased to current master.

**Comment by vszakats** [comment]
Slightly off, but the Travis CI queue may be eased a bit by enabling both 'Auto cancel' options on https://travis-ci.org/curl/curl/settings. This would drop any pending commits from the queue if a newer commits lands.

There is a similar option on AppVeyor CI, called 'Rolling builds', which could be enabled for a similar effect, but in this case, by default it will even cancel _running_ builds on the same branch on any new commit (this can be disabled to work like Travis CI.)

With these set, there would be less strain on the CI servers and the the queue could finish faster.

What do you think about it?

**Comment by jay** [comment]
Thanks. I landed this with a [slight change](https://github.com/curl/curl/blob/b7b4dc0/lib/ssh.c#L2945-L2950) so that the error message will be shown even if the api isn't available. Also note infof should end in a linefeed
~~~c
  if(data->set.ssh_compression) {
#if LIBSSH2_VERSION_NUM >= 0x010208
    if(libssh2_session_flag(ssh->ssh_session, LIBSSH2_FLAG_COMPRESS, 1) < 0)
#endif
      infof(data, "Failed to enable compression for ssh session\n");
  }
~~~

**Comment by vszakats** [comment]
Good idea, thank you @jay! (Noted the `infof()` issue &ndash; what happened is that I copied that line from the preceding `failf()` which doesn't need the `\n`)

**Pull request closed without merging**